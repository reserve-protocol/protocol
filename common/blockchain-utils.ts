import { BigNumber } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

// getChainId: Returns current chain Id
export const getChainId = async (hre: HardhatRuntimeEnvironment) => {
  let _chainId
  try {
    _chainId = await hre.network.provider.send('eth_chainId')
  } catch (e) {
    console.log('failed to get chainId, falling back on net_version...')
    _chainId = await hre.network.provider.send('net_version')
  }

  if (!_chainId) {
    throw new Error(`could not get chainId from network`)
  }
  if (_chainId.startsWith('0x')) {
    _chainId = BigNumber.from(_chainId).toString()
  }
  return _chainId
}
