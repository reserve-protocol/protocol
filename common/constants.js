const { BigNumber } = require("ethers");

const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';

const ONE_ETH= BigNumber.from("1000000000000000000");

const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1);
const MAX_UINT16 = (2**16) - 1;

const SCALE_FACTOR = 1e18;
const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString());

module.exports = {
    ZERO_ADDRESS,
    MAX_UINT256,
    MAX_UINT16,
    SCALE_FACTOR,
    BN_SCALE_FACTOR,
    ONE_ETH
};
