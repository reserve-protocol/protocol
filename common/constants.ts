import { ethers } from 'hardhat'
import { BigNumber } from 'ethers'

export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
export const ONE_ADDRESS = '0x0000000000000000000000000000000000000001'

export const ONE_ETH = BigNumber.from('1000000000000000000')

export const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1)
export const MAX_UINT192 = BigNumber.from(2).pow(192).sub(1)
export const MAX_UINT96 = BigNumber.from(2).pow(96).sub(1)
export const MAX_UINT32 = 2 ** 32 - 1
export const MAX_UINT16 = 2 ** 16 - 1

export const SCALE_DECIMALS = 18
export const SCALE_FACTOR = 10 ** SCALE_DECIMALS
export const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString())

// @dev Must match `IAsset.CollateralStatus`.
export enum CollateralStatus {
  SOUND,
  IFFY,
  DISABLED,
}

// @dev Must match `Governance.ProposalState`.
export enum ProposalState {
  Pending,
  Active,
  Canceled,
  Defeated,
  Succeeded,
  Queued,
  Expired,
  Executed,
}

// @dev Must match `Fixed.RoundingApproach`.
export enum RoundingMode {
  FLOOR,
  ROUND,
  CEIL,
}

// @dev Must match `GnosisTrade.TradeStatus`.
export enum TradeStatus {
  NOT_STARTED,
  OPEN,
  CLOSED,
}

export const FURNACE_DEST = '0x0000000000000000000000000000000000000001'
export const STRSR_DEST = '0x0000000000000000000000000000000000000002'

export const QUEUE_START = '0x0000000000000000000000000000000000000000000000000000000000000001'

// Auth roles
export const OWNER = ethers.utils.formatBytes32String('OWNER')
export const SHORT_FREEZER = ethers.utils.formatBytes32String('SHORT_FREEZER')
export const LONG_FREEZER = ethers.utils.formatBytes32String('LONG_FREEZER')
export const PAUSER = ethers.utils.formatBytes32String('PAUSER')
