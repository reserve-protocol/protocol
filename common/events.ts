import { expect } from 'chai'
import { BigNumber, ContractReceipt, Event } from 'ethers'

// TODO: Proper typing
const contains = (args: { [key: string]: any | undefined }, key: string, value: any): any => {
  expect(key in args).to.equal(true, `Event argument '${key}' not found`)

  if (value === null) {
    expect(args[key]).to.equal(null, `expected event argument '${key}' to be null but got ${args[key]}`)
  } else if (BigNumber.isBigNumber(args[key]) || BigNumber.isBigNumber(value)) {
    const actual = BigNumber.isBigNumber(args[key]) ? args[key].toString() : args[key]
    const expected = BigNumber.isBigNumber(value) ? value.toString() : value

    expect(args[key]).to.equal(value, `expected event argument '${key}' to have value ${expected} but got ${actual}`)
  } else {
    expect(args[key]).to.be.deep.equal(
      value,
      `expected event argument '${key}' to have value ${value} but got ${args[key]}`
    )
  }
}

// TODO: Proper typing for "eventArgs"
export const expectInReceipt = (receipt: ContractReceipt, eventName: string, eventArgs = {}): any => {
  if (receipt.events == undefined) {
    throw new Error('No events found in receipt')
  }

  const events = receipt.events.filter((e: Event) => e.event === eventName)
  expect(events.length > 0).to.equal(true, `No '${eventName}' events found`)

  const exceptions: string[] = []
  const event = events.find(function (e: Event) {
    for (const [k, v] of Object.entries(eventArgs)) {
      try {
        if (e.args == undefined) {
          throw new Error('Event has no arguments')
        }

        contains(e.args, k, v)
      } catch (error) {
        exceptions.push(error as string)
        return false
      }
    }
    return true
  })

  if (event === undefined) {
    // Each event entry may have failed to match for different reasons,
    // throw the first one
    throw exceptions[0]
  }

  return event
}
