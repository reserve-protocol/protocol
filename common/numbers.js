const { BigNumber } = require("ethers");
const { BN_SCALE_FACTOR } = require("./constants");

const bn = (x) => {
    if (BigNumber.isBigNumber(x)) return x;
    const stringified = parseScientific(x.toString());
    const integer = stringified.split('.')[0];
    return BigNumber.from(integer);
};

const fp = (x) => {
    return BN_SCALE_FACTOR.mul(x);
}
 
function parseScientific(num) {
    // If the number is not in scientific notation return it as it is
    if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num;

    // Remove the sign
    const numberSign = Math.sign(Number(num));
    num = Math.abs(Number(num)).toString();

    // Parse into coefficient and exponent
    const [coefficient, exponent] = num.toLowerCase().split('e');
    let zeros = Math.abs(Number(exponent));
    const exponentSign = Math.sign(Number(exponent));
    const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.');

    if (exponentSign === -1) {
        zeros -= integer.length;
        num =
            zeros < 0
                ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
                : '0.' + '0'.repeat(zeros) + integer + decimals;
    } else {
        if (decimals) zeros -= decimals.length;
        num =
            zeros < 0
                ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
                : integer + decimals + '0'.repeat(zeros);
    }

    return numberSign < 0 ? '-' + num : num;
}

module.exports = {
    bn,
    fp
}