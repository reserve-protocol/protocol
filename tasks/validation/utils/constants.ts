import { networkConfig } from '#/common/configuration'

export const whales: { [key: string]: string } = {
  [networkConfig['1'].tokens.USDT!.toLowerCase()]: '0x47ac0Fb4F2D84898e4D9E7b4DaB3C24507a6D503',
  [networkConfig['1'].tokens.USDC!.toLowerCase()]: '0xAFAaDfa18D9d63d09F19a5445e29CEc601054C5e',
  [networkConfig['1'].tokens.pyUSD!.toLowerCase()]: '0xA5588F7cdf560811710A2D82D3C9c99769DB1Dcb',
  [networkConfig['1'].tokens.RSR!.toLowerCase()]: '0x6bab6EB87Aa5a1e4A8310C73bDAAA8A5dAAd81C1',
  [networkConfig['1'].tokens.cUSDT!.toLowerCase()]: '0xb99CC7e10Fe0Acc68C50C7829F473d81e23249cc',
  [networkConfig['1'].tokens.aUSDT!.toLowerCase()]: '0x0B6B712B0f3998961Cd3109341b00c905b16124A',
  ['0x21fe646D1Ed0733336F2D4d9b2FE67790a6099D9'.toLowerCase()]:
    '0x5754284f345afc66a98fbB0a0Afe71e0F007B949', // saUSDT
  ['0x4Be33630F92661afD646081BC29079A38b879aA0'.toLowerCase()]:
    '0x5754284f345afc66a98fbB0a0Afe71e0F007B949', // cUSDTVault

  [networkConfig['1'].tokens.aUSDC!.toLowerCase()]: '0x777777c9898D384F785Ee44Acfe945efDFf5f3E0',
  [networkConfig['1'].tokens.cUSDC!.toLowerCase()]: '0x97D868b5C2937355Bf89C5E5463d52016240fE86',
  [networkConfig['1'].tokens.cUSDCv3!.toLowerCase()]: '0x7f714b13249BeD8fdE2ef3FBDfB18Ed525544B03',
  ['0x60C384e226b120d93f3e0F4C502957b2B9C32B15'.toLowerCase()]:
    '0x51eDF02152EBfb338e03E30d65C15fBf06cc9ECC', // saUSDC
  ['0xf579F9885f1AEa0d3F8bE0F18AfED28c92a43022'.toLowerCase()]:
    '0x51eDF02152EBfb338e03E30d65C15fBf06cc9ECC', // cUSDCVault
  [networkConfig['1'].tokens.RSR!.toLowerCase()]: '0x6bab6EB87Aa5a1e4A8310C73bDAAA8A5dAAd81C1',
  [networkConfig['1'].tokens.WBTC!.toLowerCase()]: '0x8eb8a3b98659cce290402893d0123abb75e3ab28',
  [networkConfig['1'].tokens.stETH!.toLowerCase()]: '0x176F3DAb24a159341c0509bB36B833E7fdd0a132',
  [networkConfig['1'].tokens.WETH!.toLowerCase()]: '0x8EB8a3b98659Cce290402893d0123abb75E3ab28',
  [networkConfig['1'].tokens.DAI!.toLowerCase()]: '0x8EB8a3b98659Cce290402893d0123abb75E3ab28',
  [networkConfig['1'].tokens.CRV!.toLowerCase()]: '0xf977814e90da44bfa03b6295a0616a897441acec',
  ['0xAEda92e6A3B1028edc139A4ae56Ec881f3064D4F'.toLowerCase()]:
    '0x8605dc0C339a2e7e85EEA043bD29d42DA2c6D784', // cvxeUSDFRAXBP LP token
  [networkConfig['1'].tokens.sDAI!.toLowerCase()]: '0x4aa42145Aa6Ebf72e164C9bBC74fbD3788045016',
  ['0xa0d69e286b938e21cbf7e51d71f6a4c8918f482f'.toLowerCase()]:
    '0x3154Cf16ccdb4C6d922629664174b904d80F2C35', // eUSD
  ['0xacdf0dba4b9839b96221a8487e9ca660a48212be'.toLowerCase()]:
    '0x8a8434A5952aC2CF4927bbEa3ace255c6dd165CD', // hyUSD
}

export const collateralToUnderlying: { [key: string]: string } = {
  ['0x21fe646D1Ed0733336F2D4d9b2FE67790a6099D9'.toLowerCase()]:
    networkConfig['1'].tokens.USDT!.toLowerCase(),
  ['0x60C384e226b120d93f3e0F4C502957b2B9C32B15'.toLowerCase()]:
    networkConfig['1'].tokens.USDC!.toLowerCase(),
  [networkConfig['1'].tokens.cUSDT!.toLowerCase()]: networkConfig['1'].tokens.USDT!.toLowerCase(),
  [networkConfig['1'].tokens.cUSDC!.toLowerCase()]: networkConfig['1'].tokens.USDC!.toLowerCase(),
  [networkConfig['1'].tokens.saEthUSDC!.toLowerCase()]:
    networkConfig['1'].tokens.aEthUSDC!.toLowerCase(),
}

export interface RTokenDeployment {
  rToken: string
  governor: string
  timelock: string
}

export const MAINNET_DEPLOYMENTS: RTokenDeployment[] = [
  {
    rToken: '0xA0d69E286B938e21CBf7E51D71F6A4c8918f482F', // eUSD
    governor: '0x7e880d8bD9c9612D6A9759F96aCD23df4A4650E6',
    timelock: '0xc8Ee187A5e5c9dC9b42414Ddf861FFc615446a2c',
  },
  {
    rToken: '0xE72B141DF173b999AE7c1aDcbF60Cc9833Ce56a8', // ETH+
    governor: '0x239cDcBE174B4728c870A24F77540dAB3dC5F981',
    timelock: '0x5f4A10aE2fF68bE3cdA7d7FB432b10C6BFA6457B',
  },
  {
    rToken: '0xaCdf0DBA4B9839b96221a8487e9ca660a48212be', // hyUSD (mainnet)
    governor: '0x22d7937438b4bBf02f6cA55E3831ABB94Bd0b6f1',
    timelock: '0x624f9f076ED42ba3B37C3011dC5a1761C2209E1C',
  },
  {
    rToken: '0xFc0B1EEf20e4c68B3DCF36c4537Cfa7Ce46CA70b', // USDC+
    governor: '0xc837C557071D604bCb1058c8c4891ddBe8FDD630',
    timelock: '0x6C957417cB6DF6e821eec8555DEE8b116C291999',
  },
  {
    rToken: '0x0d86883FAf4FfD7aEb116390af37746F45b6f378', // USD3
    governor: '0x020CB71181008369C388CaAEE98b0E69f8F4C471',
    timelock: '0xE0289984F709fc7150E646B672bfaDC879a15f14',
  },
  {
    rToken: '0x78da5799CF427Fee11e9996982F4150eCe7a99A7', // rgUSD
    governor: '0x409bAc94c4207C6627EA5f4E4FFB7128e8F654Fc',
    timelock: '0x9aD9E73e38c8506a664A3A37e8A9CE910B6FBeb4',
  },
]

export const BASE_DEPLOYMENTS: RTokenDeployment[] = [
  {
    rToken: '0xCc7FF230365bD730eE4B352cC2492CEdAC49383e', // hyUSD (base)
    governor: '0xc8e63d3501A246fa1ddBAbe4ad0B50e9d32aA8bb',
    timelock: '0xf093d7f00f3dCe6d415Be564f41Cb4bc032fb367',
  },
  {
    rToken: '0xCb327b99fF831bF8223cCEd12B1338FF3aA322Ff', // bsdETH
    governor: '0xB05C6a7242595f2E23CC6a0aB20699d63D0939Fd',
    timelock: '0x321f7493B8B675dFfE2570Bd0F164237D445b9E8',
  },
  {
    rToken: '0xfE0D6D83033e313691E96909d2188C150b834285', // iUSDC - Assets skipped (USDbC)
    governor: '0xfe637F7D5B848392c19052631d68F8AC859F71cF',
    timelock: '0xd18ED37CA912bbf1EDE93d27459d03DC4343dea1',
  },
  {
    rToken: '0xC9a3e2B3064c1c0546D3D0edc0A748E9f93Cf18d', // Vaya - Assets skipped (USDbC)
    governor: '0xEb583EA06501f92E994C353aD2741A35582987aA',
    timelock: '0xeE3eC997A37e661a42673D7A489Fbf0E5ed0C223',
  },
  {
    rToken: '0x641B0453487C9D14c5df96d45a481ef1dc84e31f', // MAAT
    governor: '0x0f7f1442dA7F687BB877Fbee0539FA8D6e4d1a02',
    timelock: '0xE67cEb03EfdF9B3fb5C3FeBF3103e2efd3a76A1b',
  },
]
