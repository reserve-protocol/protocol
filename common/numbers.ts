import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

export const bn = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) return x
  const stringified = parseScientific(x.toString())
  const integer = stringified.split('.')[0]
  return BigNumber.from(integer)
}

export const pow10 = (exponent: number): BigNumber => {
  return bn(10).pow(exponent)
}

export const fp = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) {
    return BN_SCALE_FACTOR.mul(x)
  }
  return ethers.utils.parseUnits(x.toString(), SCALE_DECIMALS)
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

function parseScientific(num: string) {
  // If the number is not in scientific notation return it as it is
  if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num

  // Remove the sign
  const numberSign = Math.sign(Number(num))
  num = Math.abs(Number(num)).toString()

  // Parse into coefficient and exponent
  const [coefficient, exponent] = num.toLowerCase().split('e')
  let zeros = Math.abs(Number(exponent))
  const exponentSign = Math.sign(Number(exponent))
  const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.')

  if (exponentSign === -1) {
    zeros -= integer.length
    num =
      zeros < 0
        ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
        : '0.' + '0'.repeat(zeros) + integer + decimals
  } else {
    if (decimals) zeros -= decimals.length
    num =
      zeros < 0
        ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
        : integer + decimals + '0'.repeat(zeros)
  }

  return numberSign < 0 ? '-' + num : num
}
