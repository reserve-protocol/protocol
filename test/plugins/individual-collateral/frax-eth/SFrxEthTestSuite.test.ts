import collateralTests from '../collateralTests'
import { CollateralFixtureContext, CollateralOpts, MintCollateralFunc } from '../pluginTestTypes'
import { resetFork, mintSfrxETH, mintFrxETH } from './helpers'
import hre, { ethers } from 'hardhat'
import { expect } from 'chai'
import { ContractFactory, BigNumberish, BigNumber } from 'ethers'
import {
  ERC20Mock,
  MockV3Aggregator,
  MockV3Aggregator__factory,
  SfraxEthMock,
  TestICollateral,
  IsfrxEth,
} from '../../../../typechain'
import { bn, fp } from '../../../../common/numbers'
import { CollateralStatus } from '../../../../common/constants'
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import {
  PRICE_TIMEOUT,
  ORACLE_ERROR,
  ORACLE_TIMEOUT,
  MAX_TRADE_VOL,
  DEFAULT_THRESHOLD,
  DELAY_UNTIL_DEFAULT,
  WETH,
  FRX_ETH,
  SFRX_ETH,
  ETH_USD_PRICE_FEED,
} from './constants'
import {
  advanceTime,
  setNextBlockTimestamp,
  getLatestBlockTimestamp,
  advanceBlocks,
} from '../../../utils/time'

/*
  Define interfaces
*/

interface SFrxEthCollateralFixtureContext extends CollateralFixtureContext {
  frxEth: ERC20Mock
  sfrxEth: IsfrxEth
}

/*
  Define deployment functions
*/

export const defaultRethCollateralOpts: CollateralOpts = {
  erc20: SFRX_ETH,
  targetName: ethers.utils.formatBytes32String('ETH'),
  rewardERC20: WETH,
  priceTimeout: PRICE_TIMEOUT,
  chainlinkFeed: ETH_USD_PRICE_FEED,
  oracleTimeout: ORACLE_TIMEOUT,
  oracleError: ORACLE_ERROR,
  maxTradeVolume: MAX_TRADE_VOL,
  defaultThreshold: DEFAULT_THRESHOLD,
  delayUntilDefault: DELAY_UNTIL_DEFAULT,
  revenueHiding: fp('0'),
}

export const deployCollateral = async (opts: CollateralOpts = {}): Promise<TestICollateral> => {
  opts = { ...defaultRethCollateralOpts, ...opts }

  const SFraxEthCollateralFactory: ContractFactory = await ethers.getContractFactory(
    'SFraxEthCollateral'
  )

  const collateral = <TestICollateral>await SFraxEthCollateralFactory.deploy(
    {
      erc20: opts.erc20,
      targetName: opts.targetName,
      priceTimeout: opts.priceTimeout,
      chainlinkFeed: opts.chainlinkFeed,
      oracleError: opts.oracleError,
      oracleTimeout: opts.oracleTimeout,
      maxTradeVolume: opts.maxTradeVolume,
      defaultThreshold: opts.defaultThreshold,
      delayUntilDefault: opts.delayUntilDefault,
    },
    opts.revenueHiding,
    { gasLimit: 2000000000 }
  )
  await collateral.deployed()
  // sometimes we are trying to test a negative test case and we want this to fail silently
  // fortunately this syntax fails silently because our tools are terrible
  await expect(collateral.refresh())

  return collateral
}

const chainlinkDefaultAnswer = bn('1600e8')

type Fixture<T> = () => Promise<T>

const makeCollateralFixtureContext = (
  alice: SignerWithAddress,
  opts: CollateralOpts = {}
): Fixture<SFrxEthCollateralFixtureContext> => {
  const collateralOpts = { ...defaultRethCollateralOpts, ...opts }

  const makeCollateralFixtureContext = async () => {
    const MockV3AggregatorFactory = <MockV3Aggregator__factory>(
      await ethers.getContractFactory('MockV3Aggregator')
    )

    const chainlinkFeed = <MockV3Aggregator>(
      await MockV3AggregatorFactory.deploy(8, chainlinkDefaultAnswer)
    )
    collateralOpts.chainlinkFeed = chainlinkFeed.address

    const frxEth = (await ethers.getContractAt('ERC20Mock', FRX_ETH)) as ERC20Mock
    const sfrxEth = (await ethers.getContractAt('IsfrxEth', SFRX_ETH)) as IsfrxEth
    const collateral = await deployCollateral(collateralOpts)

    return {
      alice,
      collateral,
      chainlinkFeed,
      frxEth,
      sfrxEth,
      tok: sfrxEth,
    }
  }

  return makeCollateralFixtureContext
}

/*
  Define helper functions
*/

const mintCollateralTo: MintCollateralFunc<SFrxEthCollateralFixtureContext> = async (
  ctx: SFrxEthCollateralFixtureContext,
  amount: BigNumberish,
  user: SignerWithAddress,
  recipient: string
) => {
  await mintSfrxETH(ctx.sfrxEth, user, amount, recipient, ctx.chainlinkFeed)
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const reduceTargetPerRef = async () => {}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const increaseTargetPerRef = async () => {}

// prettier-ignore
const reduceRefPerTok = async () => {
    await hre.network.provider.send('evm_mine', [])
}

// prettier-ignore
const increaseRefPerTok = async (
  ctx: SFrxEthCollateralFixtureContext,
  pctIncrease: BigNumberish 
) => {
  const currentBal = await ctx.frxEth.balanceOf(ctx.sfrxEth.address)
  const addBal = currentBal.mul(pctIncrease).div(100)
  await mintFrxETH(ctx.frxEth, ctx.alice!, addBal, ctx.sfrxEth.address)
  const rewardCycleEnd = await ctx.sfrxEth.rewardsCycleEnd()
  const nextTimestamp = await getLatestBlockTimestamp()
  if (nextTimestamp < rewardCycleEnd) {
    await setNextBlockTimestamp(rewardCycleEnd + 1)
    await hre.network.provider.send('evm_mine', [])
  }
  await ctx.sfrxEth.syncRewards()
  await advanceBlocks(1200 / 12)
  await advanceTime(1200)
  // push chainlink oracle forward so that tryPrice() still works
  const lastAnswer = await ctx.chainlinkFeed.latestAnswer()
  await ctx.chainlinkFeed.updateAnswer(lastAnswer)
}

const getExpectedPrice = async (ctx: SFrxEthCollateralFixtureContext): Promise<BigNumber> => {
  // Peg Feed
  const clData = await ctx.chainlinkFeed.latestRoundData()
  const clDecimals = await ctx.chainlinkFeed.decimals()

  const refPerTok = await ctx.collateral.refPerTok()
  const expectedPegPrice = clData.answer.mul(bn(10).pow(18 - clDecimals))
  return expectedPegPrice.mul(refPerTok).div(fp('1'))
}

/*
  Define collateral-specific tests
*/

// eslint-disable-next-line @typescript-eslint/no-empty-function
const collateralSpecificConstructorTests = () => {}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const collateralSpecificStatusTests = () => {
  it('does revenue hiding', async () => {
    const MockFactory = await ethers.getContractFactory('SfraxEthMock')
    const erc20 = (await MockFactory.deploy()) as SfraxEthMock
    let currentPPS = await (await ethers.getContractAt('IsfrxEth', SFRX_ETH)).pricePerShare()
    currentPPS = currentPPS.sub(currentPPS.div(1000)) // backoff slightly
    await erc20.setPricePerShare(currentPPS)

    const chainlinkFeed = <MockV3Aggregator>(
      await (await ethers.getContractFactory('MockV3Aggregator')).deploy(8, chainlinkDefaultAnswer)
    )
    const collateral = await deployCollateral({
      erc20: erc20.address,
      revenueHiding: fp('0.01'),
      chainlinkFeed: chainlinkFeed.address,
    })

    // Should remain SOUND after a 1% decrease
    await erc20.setPricePerShare(currentPPS.sub(currentPPS.div(100)))
    await collateral.refresh()
    expect(await collateral.status()).to.equal(CollateralStatus.SOUND)

    // Should become DISABLED if drops more than that
    await erc20.setPricePerShare(currentPPS.sub(currentPPS.div(99)))
    await collateral.refresh()
    expect(await collateral.status()).to.equal(CollateralStatus.DISABLED)
  })
}

// eslint-disable-next-line @typescript-eslint/no-empty-function
const beforeEachRewardsTest = async () => {}

/*
  Run the test suite
*/

const opts = {
  deployCollateral,
  collateralSpecificConstructorTests,
  collateralSpecificStatusTests,
  beforeEachRewardsTest,
  makeCollateralFixtureContext,
  mintCollateralTo,
  reduceTargetPerRef,
  increaseTargetPerRef,
  reduceRefPerTok,
  increaseRefPerTok,
  getExpectedPrice,
  itClaimsRewards: it.skip,
  itChecksTargetPerRefDefault: it.skip,
  itChecksRefPerTokDefault: it.skip,
  itChecksPriceChanges: it,
  itHasRevenueHiding: it.skip, // implemnted in this file
  resetFork,
  collateralName: 'SFraxEthCollateral',
  chainlinkDefaultAnswer,
}

collateralTests(opts)
