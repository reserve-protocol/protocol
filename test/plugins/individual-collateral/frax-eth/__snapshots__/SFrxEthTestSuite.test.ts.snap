// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting ERC20 transfer 1`] = `51304`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting ERC20 transfer 2`] = `34204`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting refresh() after full price timeout 1`] = `58984`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting refresh() after full price timeout 2`] = `54247`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting refresh() after oracle timeout 1`] = `59684`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting refresh() after oracle timeout 2`] = `58015`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting refresh() during SOUND 1`] = `73809`;

exports[`Collateral: SFraxEthCollateral collateral functionality Gas Reporting refresh() during SOUND 2`] = `73809`;
